Fix multiple issues with one-way collisions

madmiraal opened the pull request:
This is a salvage of #36280, because, despite the original author's [decision](https://github.com/godotengine/godot/pull/36280#issuecomment-702923451) to close #36280 in favour of #38471, #38471 does not solve all the issues that were solved by #36280.

This PR does a few things:
- For RigidBodies, uses the collision normal determined by relative motion to determine whether or not a one-way collision has occurred.
- For KinematicBodies, performs additional checks to ensure a one-way collision has occurred, and averages the recovery step over all collision shapes.

I have made some slight changes to the original PR to minimise the necessary changes and make this PR easy to review. I have also made the original author a co-author of the commit.

I have tested and confirmed that this PR fixes the following issues:
Fixes #25732
Fixes #25967
Fixes #28794
Fixes #34242
Fixes #35776
Fixes #40006
Fixes #42175
Fixes #43266

Note: #36280 claimed to fix other issues too. However, I found that several of these issues were already fixed (so I have updated them accordingly), or are actually not fixed by this PR.

**Edit:** Updated to remove the change that caused the bug [reported](https://github.com/godotengine/godot/pull/42574#issuecomment-711069694) by @Rhathe.

**Edit:** Updated to fix the bugs identified by @Rhathe [here](https://github.com/godotengine/godot/pull/42574#issuecomment-715392217) and [here](https://github.com/godotengine/godot/pull/42574#issuecomment-716117102).

**Edit:** Updated to fix #43266, which is the same bug identified by @AttackButton [here](https://github.com/godotengine/godot/pull/42574#issuecomment-719957095).

madmiraal committed:
Fix multiple issues with one-way collisions

For RigidBodies, uses the collision normal determined by relative motion
to determine whether or not a one-way collision has occurred.

For KinematicBodies, performs additional checks to ensure a one-way
collision has occurred, and averages the recovery step over all collision
shapes.

Co-authored-by:    Sergej Gureev <sergej.gureev@relex.fi>

madmiraal commented:
As suggested by @capnm [here](https://github.com/godotengine/godot/pull/42575#discussion_r500077601), updated to ensure a zero vector is not normalized.

bemyak commented:
> #38471 does not solve all the issues that were solved by #36280

Did you test it with or without patch from [this comment](https://github.com/godotengine/godot/pull/38471#issuecomment-696539734) ?

RandomShaper commented:
I've checked the changes carefully. I don't see anything fishy and I understand the rationale of most of them.

I've also tested the 3.2 version of this PR on my own game, which doesn't use one-way much, but also as a way to spot potential regressions or relevant behavior changes in the general behavior of body collisions. Everything seems alright.

So I'm already giving my approval, but I'd like to see a project with heavy use of one-way collisions running with this changes, to confirm for good that the changes work as expected.

Rhathe commented:
Last time I tested these changes I noticed it introduced a bug in the 3.2 branch as I noted here: https://github.com/godotengine/godot/pull/36280#issuecomment-623042390

This is the original:
![original](https://user-images.githubusercontent.com/1882199/80896854-d433de80-8cc0-11ea-9d7e-bc5b55d9c7e3.gif)

This is with the changes:
![fix](https://user-images.githubusercontent.com/1882199/80896863-e6ae1800-8cc0-11ea-99ce-71bb6ff12eaa.gif)

Is that bug still here?

[test_move_one_way.zip](https://github.com/godotengine/godot/files/4569491/test_move_one_way.zip)

capnm commented:
I think this is a bug in your project, if you reset the excessive `one_way_collision_margin`
it works as expected:
![grafik](https://user-images.githubusercontent.com/4047289/96355640-0953b200-10e4-11eb-9919-543104af8e95.png)

Rhathe commented:
> I think this is a bug in your project, if you reset the excessive [`one_way_collision_margin`](https://docs.godotengine.org/en/stable/classes/class_collisionshape2d.html#class-collisionshape2d-property-one-way-collision-margin)
> it works as expected:
> ![grafik](https://user-images.githubusercontent.com/4047289/96355640-0953b200-10e4-11eb-9919-543104af8e95.png)
> (You are probably trying to abuse this margin for »side_way« collisions.)

The excessive [`one_way_collision_margin`] is only to make sure this effect shows up consistently and thus a more obvious bug to fix. When set with just a value of 1, the bug still shows up, you may just need to run it a couple of times to see it occur.

Even if the bug only showed up with an "excessive" value, it is still a bug either way that does not show up before this change.

capnm commented:
> The excessive one_way_collision_margin is only to make sure this effect shows up consistently and thus a more obvious bug to fix.

Please fill then an issue for that what you think is »obvious bug to fix«.
Thanks.

Rhathe commented:
> > The excessive one_way_collision_margin is only to make sure this effect shows up consistently and thus a more obvious bug to fix.
> 
> Please fill then an issue for that what you think is »obvious bug to fix«. I can't reproduce it.
> (The one_way_collision_margin should be probably removed, as the one way collision now works even by high speed without it.)
> Thanks.

I'm not sure an issue should be created for a bug introduced in a pull request that has not been merged yet. I've added a new version of the test project so that it more easily shows up. The only changes are that the KinematicBody2D moves with a vertical velocity of 3000 instead of 1000, and the `one_way_collision_margin` is set to 0. The bug now should show up all the time.

[test_move_one_way_2.zip](https://github.com/godotengine/godot/files/5397090/test_move_one_way_2.zip)

capnm commented:
> the KinematicBody2D moves with a vertical velocity of 3000 instead of 1000

I don't think this is a reasonable speed, indeed now the kinematic body jumps at the one way collision
shape.

madmiraal commented:
Updated, so a perpendicular collision is detected as a valid collision. Now both the `RigidBody` and the `KinematicBody` behave the same way:
![42574-fixed](https://user-images.githubusercontent.com/9253928/96361182-8a7a6b80-111b-11eb-9302-4d06af481041.gif)

capnm commented:
[pr42574-move-horizontal2.zip](https://github.com/godotengine/godot/files/5397281/pr42574-move-horizontal2.zip)

![Bildschirmfoto von 2020-10-18 11-22-14](https://user-images.githubusercontent.com/4047289/96363407-7d1dab00-1134-11eb-8d44-5837b4189c12.png)

madmiraal commented:
@capnm The `RigidBody` is not moving perfectly horizontally. It's actually rocking slightly, which is why you see the contact points flickering. This will result in slightly different collisions due to slightly different starting conditions. With the original example project, the collision normal is `<-1, 0.000295927>` i.e. pointing slightly upwards; so it's stopped. In your example project, the collision normal is `<-1, -0.000422864>` i.e. pointing slightly downwards; so it's allowed through. Setting the `RigidBody` to `Character` mode, prevents rotations i.e. the rocking and the collision normal is `<-1, 0>`, which is stopped as with the `KinematicBody`

Note: `RigidBody` one-way collisions are only checked at the time of collision. The `RigidBody` is then either allowed or disallowed to pass through until it exits the one-way collision body. This results in the differences seen around the edge case of a perpendicular collision.

Rhathe commented:
I'd say it's the RigidBody2D behavior that needs to be fixed instead of the KinematicBody2D's behavior. That collision check is for if the collision vector is perpendicular to the one way vector, so it results in colliding with the side if your character jumps and falls to the side of a one way platform. I uploaded another version of the test code to show that, this time with both bodies being KinematicBody2Ds. I also show that the bug still shows up if the ground platform is tilted by -1 degrees.

![test_move_one_way_3](https://user-images.githubusercontent.com/1882199/96374582-8e20e900-1141-11eb-94f5-74bc3b3624c4.gif)

[test_move_one_way_3.zip](https://github.com/godotengine/godot/files/5397854/test_move_one_way_3.zip)

bemyak commented:
> Updated, so a perpendicular collision is detected as a valid collision

@madmiraal , I don't think this is the way one-way collisions are supposed to work. As the name suggests, it should collide only in one direction. Many platformers rely on this: if a character didn't make high enough jump to reach the top of the platform, if should just pass through instead of hitting an invisible wall.

Expected trajectory:
![Screenshot_20201018_202630](https://user-images.githubusercontent.com/6333105/96375317-03f87500-1181-11eb-85bc-624076f43a8c.png)

 
With horizontal collision:
![Screenshot_20201018_202852](https://user-images.githubusercontent.com/6333105/96375297-d9a6b780-1180-11eb-8b30-43932756ef7b.png)

This becomes much more worse when using tile maps with one-way collisions.

madmiraal commented:
Makes sense. I'll update it so that both `KinematicBody` and `RigidBody` treat perpendicular contacts as pass through. I'll also have a look at the jumping `KinematicBody` issue.

madmiraal commented:
Updated to ensure that both `KinematicBody` and `RigidBody` treat perpendicular contacts as pass through; removed the change the was causing the bug; and confirmed that all the fixed issues are still fixed.
![42574-fixed1](https://user-images.githubusercontent.com/9253928/96478388-8b042680-122f-11eb-9be2-3fd97076c2f8.gif)
![42574-fixed2](https://user-images.githubusercontent.com/9253928/96478411-8f304400-122f-11eb-83e7-c69a17d265a2.gif)

akien-mga commented:
To validate this PR and potential future changes needed to one-way collisions, it would be great to have a test project that uses one-way collisions in all the different ways it can appear in user games (or at least as exhaustive as we can think of), so we can test before/after on this and future PRs, and improve this integration test project as new issues are reported/fixed.

Would anyone be interested in doing this? Note that it doesn't have to include *all* use cases from the start, we can start with a best effort scene and improve it little by little.

Zireael07 commented:
@qarmin, maybe you'd be interested in providing a test project? Once upon a time it seemed your hobby was making a project that was making Godot crash...

AttackButton commented:
Ok. Using @bemyak last build and checking the one-way collision as true in the KinematicBody2D below, it ends up even passing through the staticbody2D platform.

![one-way_bug1](https://user-images.githubusercontent.com/25422687/96661257-0940f580-1322-11eb-9d2d-4fe9861a49fa.gif)

I used the same project of the https://github.com/godotengine/godot/issues/40006 bug:
[KinematicMoving.zip](https://github.com/godotengine/godot/files/5412618/KinematicMoving.zip)

I also found other bugs with the KB2D moving platforms with one-way collision as false in my game (vertical and circular ones).
Like the Player getting stuck without being able to move horizontally when the vertical platform goes up, having it's movement reestablished when the platform goes down. This does not happen in version 3.2.3 stable, the platforms work perfectly.

Another bug is that if the one-way collision is checked as true in the KB2D moving platform, the Player will pass-through it, ignoring it completely.

Thanks for the work of all of you.

madmiraal commented:
@AttackButton Please upload MRPs for the bugs you mention. This PR has been updated to remove a change that caused the bug that @Rhathe identified. The bug was associated with high speed `KinematicBodies`; so it may be the same bug you're referring to, but I can't tell without MRPs.

madmiraal commented:
In response to @akien-mga [comment](https://github.com/godotengine/godot/pull/42574#issuecomment-712850139), to help validate this PR, I've created [Godot Gym](https://github.com/madmiraal/godot-gym) with three initial tests based on the three key issues:
1. **Swimming Pool Tests:** Contains tests. It's based on the [updated MRP](https://github.com/godotengine/godot/issues/25732#issuecomment-466669002) from @DleanJeans in the reopened #25732.
2. **Block Climb Test:** This is an exaggeration of the `TileMap` issue raised by @bemyak [above](https://github.com/godotengine/godot/pull/42574#issuecomment-711318352) i.e. that collisions with the side of a one-way collision object should be treated as pass through. I think @Rhathe disagrees with this view from his comment [here](https://github.com/godotengine/godot/issues/42916#issuecomment-712936847).
3. **Corner grip tests:** This should test all the corner cases. I believe the corner cases should be treated as: "Should never collide unless: the direction is downward and it doesn't hit the side first." Again, I think @Rhathe disagrees with this view from his comment [here](https://github.com/godotengine/godot/issues/42916#issuecomment-712956424). Note: Some of the failure cases are not limited to one-way collisions and are related to #23140 i.e. `KinematicBody` zero pixel overlap collisions.

Note: This PR fixes the issues tested in test 1: Swimming Pool Tests and 2: Block Climb Test (assuming @Rhathe agrees). It does not address the corner cases identified in test 3: Corner Grip Tests. In fact the results suggest it now fails two more; so that will need to be investigated to see whether or not this PR would cause a noticeable regression.

AttackButton commented:
@madmiraal Here is the MRP:

[KinematicMoving-one-way.zip](https://github.com/godotengine/godot/files/5418651/KinematicMoving-one-way.zip)

In addition to the gifs above, even with one-way collision as false, the Player does not move to the left or to the right when the vertical platform is going up.

madmiraal commented:
@AttackButton, with the current version of this PR, I don't see the issues you describe i.e. the KinematicBody moves freely on the moving platforms.
Move and slide:
![42574-fixed4](https://user-images.githubusercontent.com/9253928/96840211-1228e880-1442-11eb-9cb3-75d3bd0a687a.gif)
Move and slide with snap:
![42574-fixed3](https://user-images.githubusercontent.com/9253928/96840284-25d44f00-1442-11eb-9480-fa6f7a92e127.gif)

AttackButton commented:
Nice. I believe it is working like the 3.2.3 stable now.

madmiraal commented:
> In fact the results suggest it now fails two more; so that will need to be investigated to see whether or not this PR would cause a noticeable regression.

I've done some digging, and I've found some interesting things. First, the Corner Grip Tests confirms that this PR fixes a number of issues:
1. A `KinematicBody` no longer catches the corner when coming from below a one-way platform i.e. resolves #25967.
2. A `KinematicBody` no longer misses the corner when coming from above.
3. A `RigidBody` no longer collides with the sides of one-way shapes.

This accounts for 5 of the 16 failed tests reported by the Corner Grip Tests under 3.2.3.stable. Another 9 are due to other, non-one-way specific issues e.g. #23140. The remaining 2 are borderline cases: two corners colliding (could be treated as either the top or the side) or colliding with the side when moving perpendicular to the side.

Second, the Corner Grip Tests with this PR have exposed an issue with the `RigidBody` collision solver. See #43014. The Corner Grip Tests report 7 new failures when run against this PR due to this issue. However, this is an existing issue that has been exposed by these tests, not a regression caused by this PR.

In addition, the Corner Grip Tests have revealed an additional problem with `KinematicBodies` glancing the corner of a normal `CollisionShape` too. See #43001 .

Rhathe commented:
Found a possible regression here. When dealing with one way polygons there are situations where the KinematicBody2D will fall right through the polygon. The steepness of the left shape portion seems to prevent it from being considered for one-way collision in both cases, but for the PR this also affects the flat shape portion.

This is the current PR:
![pr](https://user-images.githubusercontent.com/1882199/97018902-ab78fd00-151d-11eb-9896-5f79e5c2887d.gif)

This is 3.2.3 stable:
![stable](https://user-images.githubusercontent.com/1882199/97018925-b59afb80-151d-11eb-8479-54096cc8ac24.gif)

[test_move_one_way_polygon.zip](https://github.com/godotengine/godot/files/5430192/test_move_one_way_polygon.zip)

capnm commented:
> Found a possible regression here.

Works with one_way_collision_margin = 30.0 ;-)

Rhathe commented:
> > Found a possible regression here.
> 
> Works with one_way_collision_margin = 30.0 or slower y-speed ;-)

Set the velocity vector as `Vector2(30, 60 * 2)` (which is a y-speed of only 2 pixels per frame) and the position as `559, 214` and you still get the same results:

![slower](https://user-images.githubusercontent.com/1882199/97102923-d1b6ae00-167f-11eb-9757-92b2abc4690a.gif)

You can probably find even slower y-speeds that still show this bug. As for the one_way_collision_margin of 30.0, high collision margin values have their issues as detailed in #42914.

Rhathe commented:
Similar to the regression bug above, and also seems to be in part caused by https://github.com/godotengine/godot/issues/42916#issuecomment-712956424:

This is the current PR:
![pr](https://user-images.githubusercontent.com/1882199/97103206-5d7d0a00-1681-11eb-9c10-bf603aa4ddfd.gif)

This is 3.2.3 stable:
![stable](https://user-images.githubusercontent.com/1882199/97103214-640b8180-1681-11eb-82be-e086c7c1da0c.gif)

[test_move_one_way_straight_down_corner_2.zip](https://github.com/godotengine/godot/files/5434353/test_move_one_way_straight_down_corner_2.zip)

madmiraal commented:
Updated to include a fix for the bugs identified by @Rhathe [here](https://github.com/godotengine/godot/pull/42574#issuecomment-715392217) and [here](https://github.com/godotengine/godot/pull/42574#issuecomment-716117102).

Rhathe commented:
The latest update introduced a regression where entering a long one way platform from the middle of its side is producing a collision, even with a one way margin of 1.

This is the current PR:
![pr](https://user-images.githubusercontent.com/1882199/97669129-d2718a80-1a59-11eb-9bd2-ac085b8b6fbf.gif)

This is 3.2.3 stable:
![stable](https://user-images.githubusercontent.com/1882199/97669138-d7ced500-1a59-11eb-9e2b-b8b8752e0dcd.gif)

[test_move_one_way_enter.zip](https://github.com/godotengine/godot/files/5463368/test_move_one_way_enter.zip)

madmiraal commented:
Oops! My bad. I'll have a look and see what I've done.

madmiraal commented:
Updated to properly fix the bugs identified by @Rhathe. Now also fixes #43012, but no longer fixes #28895, which is related to #43014.

AttackButton commented:
Ok, I made some tests in my game, and this PR is working better in almost all cases against the 3.2.3-stable. But after I found https://github.com/godotengine/godot/issues/43229 that is a bug for Godot 3.2.3-stable and this PR as well, I found this other situation below with a CollisionPolygon2D with one-way collision as true only using this PR.

With this PR, even in the brief moments that the Player is able to stay above the CollisionPolygon2D, I believe that is the effect of the stop_on_slope, otherwise the Player would fall directly.

Godot 3.2.3-Stable:

![1_CollisionPolygon2D-3 2 3-Stable-one-way-true](https://user-images.githubusercontent.com/25422687/97784263-1e1c5500-1b7c-11eb-8b7a-78aa60c2dd3c.gif)

This PR:

![2_CollisionPolygon2D-PR-Stable-one-way-true](https://user-images.githubusercontent.com/25422687/97784344-a4d13200-1b7c-11eb-833f-3466f580d4fd.gif)

[KinematicMoving-one-way-CollisionPolygon2D.zip](https://github.com/godotengine/godot/files/5469420/KinematicMoving-one-way-CollisionPolygon2D.zip)

madmiraal commented:
Updated to fix the bug identified by @AttackButton [above](https://github.com/godotengine/godot/pull/42574#issuecomment-719957095), which is actually pre-existing bug #43266, exposed by earlier versions of this PR.

Rhathe commented:
The update reintroduces the regression from https://github.com/godotengine/godot/pull/42574#issuecomment-716117102 with a few changed starting positions.

This is the current PR:
![pr](https://user-images.githubusercontent.com/1882199/97933994-61c0bb80-1d42-11eb-83e9-d0e09f3b65b5.gif)

This is 3.2.3 stable:
![stable](https://user-images.githubusercontent.com/1882199/97934006-65544280-1d42-11eb-9a79-71bc9f0bed1f.gif)

Generally I think this bug will tend to show up if the one way collision check in `_rest_cbk_result` doesn't match the check found in `PhysicsServer2DSW::_shape_col_cbk`.

[test_move_one_way_straight_down_corner_3.zip](https://github.com/godotengine/godot/files/5478730/test_move_one_way_straight_down_corner_3.zip)

madmiraal commented:
> Generally I think this bug will tend to show up if the one way collision check in _rest_cbk_result doesn't match the check found in PhysicsServer2DSW::_shape_col_cbk.

I thought so too, but:
1. The [issue](https://github.com/godotengine/godot/pull/42574#issuecomment-719957095) identified by @AttackButton is actually due to the collision solver reporting a collision when the overlap is negative (a separate bug that I haven't had time to create an issue for yet due to the difficulty in creating an MRP). The update is a workaround to ignore small overlaps (positive or negative) as they (so far) only reveal themselves during later unsticking steps i.e. when the overlap is supposed to be zero.
2. Whether or not to ignore a one-way `CollisionShape` is calculated during the unsticking phase. It basically decides whether or not it is already inside the `CollisionShape`. Here the `valid_depth` parameter is relevant, because it needs to account for moving platforms. However, when colliding with the object for the first time, we don't want this work around to apply, because, when extracting collision information, the overlap is deliberately small and we don't want it to be ignored because it is small.

The issue identified [above](https://github.com/godotengine/godot/pull/42574#issuecomment-720806240) is both related to point 2 and #16414 (see [comment](https://github.com/godotengine/godot/issues/16414#issuecomment-687757743)). When determining the safe and unsafe distance to travel, it is actually colliding with the left-hand shape; instead of ignoring it as a one-way collision object, because the overlap is too small i.e. point 2. Then, when attempting to extract the collision information, it ignores it, because the overlap is too small. This happens whether or not the right-hand shape is a one-way collision or not i.e. #16414.

It is not visible when the right-hand shape is not a one-way shape, because `move_and_slide()` is called again in the next physics iteration. When the shape is not a one-way shape, the unsticking part of the algorithm pushes it up. When it is a one-way shape, it is deemed to already be inside the shape and so it's ignored. This problem is revealed by only performing one `move_and_slide()` step.

Finally, it's worth pointing out that the issue identified [above](https://github.com/godotengine/godot/pull/42574#issuecomment-720806240) requires the overlap between the `KinematicBody2D` and the left-hand `StaticBody2D` to be less than 0.003. i.e. this issue is not seen if the initial transform of the `KinematicBody2D` was not 631.999, but 632 (or anything less than equal to 631.996).

madmiraal commented:
Updated to ensure the work around applied in the previous update only applies during the unsticking phase.

madmiraal commented:
Updated with a small refactor to remove some unneeded code.

akien-mga commented:
Amazing work from all involved, developers and QA testers :)

I think this has reached a pretty good status to be merged now, and any potential regressions could be handled in follow up PRs. WDYT?

Rhathe commented:
One small regression found related to #25967 and #43012. Different starting values can surface it either on the current PR, 3.2.3 stable, both, or neither. These different test case versions can be found in #25967.

This is the current PR:
![pr](https://user-images.githubusercontent.com/1882199/99487117-9f713700-2933-11eb-9934-ee20a94aed8b.gif)

This is 3.2.3 stable:
![stable](https://user-images.githubusercontent.com/1882199/99487126-a39d5480-2933-11eb-830c-a6e8841fd565.gif)

[test_move_one_way_snap_on_top_pr.zip](https://github.com/godotengine/godot/files/5557946/test_move_one_way_snap_on_top_pr.zip)

madmiraal commented:
This is the separate pre-existing issue #43012, that is not specific to one-way collisions.

Rhathe commented:
It might not be specific to one-way collisions, but it is much more apparent when you use tiles for one way collision, as you would constantly encounter situations where you're moving towards the corner of one of the tiles while going through other tiles.

Also encountered another regression, this time related to 90-degree rotated one way platforms.

This is the current PR:
![pr](https://user-images.githubusercontent.com/1882199/99550156-68794080-2988-11eb-8dbb-e91bdad1b476.gif)

This is 3.2.3 stable:
![stable](https://user-images.githubusercontent.com/1882199/99550193-6fa04e80-2988-11eb-823b-3db03844052e.gif)

[test_move_one_way_rotated.zip](https://github.com/godotengine/godot/files/5561017/test_move_one_way_rotated.zip)

reduz commented:
My requirement for merging this PR has not changed, we need a proper test suite for all expected behaviors, make sure community agrees to it and have it to check for regressions.

madmiraal commented:
@reduz Your requirement will never be met, because it cannot be met. It is neither quantified ("make sure community agrees to it") nor constrained ("all expected behaviors").

While I agree that testing is important, and an open [test suite](https://github.com/madmiraal/godot-gym) is the useful, it will never be able to guarantee there will not be regressions (only reduce the likelihood). Programming is hard, physics programming especially so, but the strength of open source software is many contributors providing incremental improvements over time. Putting unrealistic demands on contributors and not providing constructive feedback will only alienate them and result in [numerous](https://github.com/godotengine/godot/issues?q=is%3Aissue+is%3Aopen+label%3Atopic%3Aphysics), [serious](https://github.com/godotengine/godot/issues?q=is%3Aissue+is%3Aopen+label%3Atopic%3Aphysics+label%3Acrash) [problems](https://github.com/godotengine/godot/issues/35780) and [regressions](https://github.com/godotengine/godot/issues/35713) [not being addressed](https://github.com/godotengine/godot-proposals/issues/570) for [years](https://github.com/godotengine/godot/issues/2092) despite [numerous](https://github.com/godotengine/godot/pulls?q=is%3Apr+is%3Aopen+label%3Atopic%3Aphysics) submitted [fixes](https://github.com/godotengine/godot/pull/35945); including this one.

Zireael07 commented:
@madmiraal: I think reduz meant tests in Godot itself, as Godot recently gained that capability.

Also the part about "community agrees to it" is... poorly worded. I can't tell what the 'it' is supposed to be - the bugfix? or the list of expected behaviors?

akien-mga commented:
This has been heavily tested by community members and @madmiraal did make a test project as requested by @reduz, so not sure what more would be expected:
https://github.com/madmiraal/godot-gym

What I would expect is a review of the code by @reduz, but excluding that, the approach I would recommend is to merge and then handle potential regressions in dedicated bug reports (and improve https://github.com/madmiraal/godot-gym along the way to test for those yet untested scenarios).

The only thing which I'm not sure yet is https://github.com/godotengine/godot/pull/42574#issuecomment-729754425 which seems to be a pre-existing bug that this PR aggravates?

vonedo3011 commented:
Physics in Godot has been bad for years. This PR provides a lot of major improvements and fixes a lot of issues in my projects, IMO Godot really really needs this. 

Hope to see this merged.

madmiraal commented:
> The only thing which I'm not sure yet is #42574 (comment) which seems to be a pre-existing bug that this PR aggravates?

This PR doesn't address existing issues with rotated objects e.g. #25313, because there is a bigger problem with rotated objects as rotating the platform (red box) a bit more shows:
<table><tr><td>3.2.3.stable

![42574i 2 3 2 3](https://user-images.githubusercontent.com/9253928/100538383-5b314100-3227-11eb-911d-525dabd16e11.gif)
</td><td>With #42574

![42574i 2 fixed](https://user-images.githubusercontent.com/9253928/100538391-66846c80-3227-11eb-960f-36d1a664d6e2.gif)
</td></tr></table>
With this PR (#42574) the behaviour appears correct.

However, a few more rotations will reveal the complexity of the problem. First lets rotate the scene, so that the platform is not rotated:
<table><tr><td>3.2.3.stable

![42574i 4 3 2 3](https://user-images.githubusercontent.com/9253928/100538544-54ef9480-3228-11eb-9faf-d7e5f2f28ec9.gif)
</td><td>With #42574

![42574i 4 fixed](https://user-images.githubusercontent.com/9253928/100538560-6173ed00-3228-11eb-8917-b928817e3197.gif)
</td></tr></table>
With this PR (#42574) the behaviour still appears correct.

However, if we now un-rotate the player:
<table><tr><td>3.2.3.stable

![42574i 5 3 2 3](https://user-images.githubusercontent.com/9253928/100538676-20c8a380-3229-11eb-9692-3dd1150baa9c.gif)
</td><td>With #42574

![42574i 5 fixed](https://user-images.githubusercontent.com/9253928/100538697-39d15480-3229-11eb-8c8e-dcf5f0141dbd.gif)
</td></tr></table>
The behaviour with this PR (#42574) changes, but, not only is it consistent with the current behaviour, it is the required, correct behaviour to ensure bodies, with simulated gravity, don't get caught inside adjacent one-way platforms. See the explanation [above](https://github.com/godotengine/godot/pull/42574#issuecomment-711318352).

Therefore, I think, as [suggested](https://github.com/godotengine/godot/pull/42574#issuecomment-725973734) by @akien-mga prior to this separate, pre-existing issue being raised, that "this has reached a pretty good status to be merged now, and any potential regressions could be handled in follow up PRs."; including discussing the appropriate behaviour with rotated platforms and bodies.

Rhathe commented:
> The only thing which I'm not sure yet is #42574 (comment) which seems to be a pre-existing bug that this PR aggravates?

The pre-existing bug had to deal with hugging the platform when colliding with corners. This bug deals with 90 degree rotated objects colliding on the bottom.

> This PR doesn't address existing issues with rotated objects e.g. #25313, because there is a bigger problem with rotated objects as rotating the platform (red box) a bit more shows:

#25313 doesn't actually deal with rotated objects. Take the minimum project from there, remove the 90 degree shape rotation, move the static bodies to the side, and give the kinematic body a horizontal velocity, and you will see the same issue. The actual problem there is that the testing kinematic body doesn't take its own one way collision into account when colliding with objects, only the other objects' one way collision.

Your cases don't really deal with rotated bodies either, as you can get the same results from having a diamond shaped object.

| 3.2.3 stable  | current PR |
| --- | --- |
| ![stable](https://user-images.githubusercontent.com/1882199/100554919-4a53f000-3266-11eb-8322-6452b422bec3.gif) | ![pr](https://user-images.githubusercontent.com/1882199/100554925-52ac2b00-3266-11eb-9276-8f71e4a296ef.gif) |

[test_move_one_way_diamond.zip](https://github.com/godotengine/godot/files/5612930/test_move_one_way_diamond.zip)

Having 90 degree rotated objects not working correctly is a significant regression, as the only way to have something like [one-way walls/doors](https://github.com/godotengine/godot/issues/20204) is to [rotate it by 90 degrees](https://github.com/godotengine/godot/issues/12791#issuecomment-343350675), as you cannot specify one way collision direction, and is essential if you want one way directions in a [top-down game](https://github.com/godotengine/godot/issues/18493). While for many rotations it seems this PR has better behavior than the current version, 90 degree rotations are an arguably more important use case and this PR would necessarily have to reopen those past issues.

robbertzzz commented:
@akien-mga I'm running into a lot problems with a project that would all be solved by this PR. Are there any chances of this getting cherry picked for 3.2.4? From what I can read in the thread the plan was to merge this about a month ago, really hoping that that's still possible for the next release!

vonedo3011 commented:
Can maybe @pouleyKetchoupp take a look at this PR? This is essential for 2D physics.

pouleyKetchoupp commented:
> 
> 
> Can maybe @pouleyKetchoupp take a look at this PR? This is essential for 2D physics.

Yes, I'm planning to spend some time next week to test and review this PR.

pouleyKetchoupp commented on servers/physics_2d/body_pair_2d_sw.cpp:
```diff
@@ -288,21 +288,17 @@ bool BodyPair2DSW::setup(real_t p_step) {
+				if (!c.reused) {
+					continue;
 				}
+				if (c.normal.dot(direction) >= 0) { //greater (normal inverted)
```
In order to get more consistent results with side collisions, the best would be to add a little threshold to this check so it's not affected by float precision around 0.

Something like:
```
#define ONE_WAY_DIR_THRESHOLD 0.001

// A case
if (c.normal.dot(direction) > -ONE_WAY_DIR_THRESHOLD) //greater (normal inverted)

// B case
if (dot < ONE_WAY_DIR_THRESHOLD) //less (normal ok)
```

Compared to the current PR's version, I'm getting similar results, and it allows rotations of 180 degrees to pass as well.

madmiraal replied:
This part of the code was not changed by this PR. Only the indenting from removing the `if` statement (line 291) above:

![Screenshot_2021-01-06_17-06-34](https://user-images.githubusercontent.com/9253928/103798548-c7327080-5041-11eb-8f18-f185af3b5cbd.png)

My [testing](https://github.com/madmiraal/godot-gym) has also shown that this is not required. Do you have an MRP where this is required?

Bear in mind that this is an edge case, and, in general, `RigidBodies` are continuously moving and rotating, so most users wouldn't notice the difference. Furthermore, as [discussed](https://github.com/godotengine/godot/pull/42574#issuecomment-735372133) above, addressing the correct behaviour for rotated bodies is more involved than just ensuring that 90° rotated bodies (which precision errors don't actually allow us to do) behave consistently.

Ultimately this should be consistent with `KinematicBody`; so unless we decide that addressing the rotated bodies issues needs to be done in this PR, this should be kept as it is.

pouleyKetchoupp replied:
The logic behind my proposed change is to make things consistent when the body hits the side of a platform. From the previous discussion, it was decided that it would always go through in this case. That could be ensured for more cases by adding a little flexibility for mathematical approximation in this check.

The case I'm talking about is not 90°, but 180° rotation (platform upside-down).

You can reproduce it in your test suite (Around the clock, tests 385 & 387).

Or in the new test I've added recently in the physics test framework, by setting a platform angle of 180°:
https://github.com/godotengine/godot-demo-projects/pull/573
This one is mostly based on your Around the Clock test, but you can set parameters more easily to reproduce certain cases.

My suggestion is a little change, it adds to consistency and doesn't seem to break anything. I think the same should apply to KinematicBody movements, but I'm just starting on that case so I can't point out the code to change yet (if any).

pouleyKetchoupp replied:
Also, you *have* changed this code from `> 0` to `>= 0` in this PR, probably for a similar reason (making side collision go through in some other cases) so it makes sense to go a bit further in that direction and ensure more cases to work this way.

pouleyKetchoupp reviewed (changes requested):
I've only tested the cases with rigid body, not kinematic body yet.

Apart from the change I've pointed out to make things more consistent with side collision, it seems like the changes are good. Results with rotated bodies are not perfectly consistent, but it's already the case without this PR so it's not really a regression and needs to be addressed separately.

I'll test and review kinematic bodies later this week.

pouleyKetchoupp commented on servers/physics_2d/physics_server_2d_sw.cpp:
```diff
@@ -149,24 +149,19 @@ void PhysicsServer2DSW::_shape_col_cbk(const Vector2 &p_point_A, const Vector2 &
-			return;
+		if (cbk->valid_depth < 10e20) {
+			if (rel_length2 > cbk->valid_depth * cbk->valid_depth ||
+					(rel_length2 > CMP_EPSILON && cbk->valid_dir.dot(rel_dir.normalized()) <= 0)) {
```
Same thing as for rigid bodies, result are more consistent when using:
`cbk->valid_dir.dot(rel_dir.normalized()) < ONE_WAY_DIR_THRESHOLD`

pouleyKetchoupp commented on servers/physics_2d/space_2d_sw.cpp:
```diff
@@ -376,25 +376,25 @@ struct _RestCallbackData2D {
+
+	Vector2 normal = contact_rel / len;
+
+	if (rd->valid_dir != Vector2() && rd->valid_dir.dot(normal) >= 0) {
```
And also here with:
`rd->valid_dir.dot(normal) > -ONE_WAY_DIR_THRESHOLD`

pouleyKetchoupp reviewed (changes requested):
After testing KinematicBody2D, I can confirm it would be all more consistent with this little margin on normal checks but I don't see anything else that really needs to be changed.

With my proposed changes in both RigidBody2D and KinematicBody2D, I'm getting the exact same results for both and almost all tests pass (there are still some failing at specific angles, but it's not worse than before this PR).

Here's what I'm getting in details using the test from https://github.com/godotengine/godot-demo-projects/pull/573:

Body angle 0, Platform size 64:
`All pass`

Body angle 45, Platform size 64:
`Test FAILED: size=64.0, angle=195.0, body angle=45.0`
`Test FAILED: size=64.0, angle=345.0, body angle=45.0`

Body angle 45, Platform size 32:
`Test FAILED: size=32.0, angle=195.0, body angle=45.0`
`Test FAILED: size=32.0, angle=210.0, body angle=45.0`
`Test FAILED: size=32.0, angle=225.0, body angle=45.0`
`Test FAILED: size=32.0, angle=330.0, body angle=45.0`
`Test FAILED: size=32.0, angle=345.0, body angle=45.0`

pouleyKetchoupp commented on servers/physics_2d/space_2d_sw.cpp:
```diff
@@ -851,7 +850,7 @@ bool Space2DSW::test_body_motion(Body2DSW *p_body, const Transform2D &p_from, co
 				Vector2 a = sr[i * 2 + 0];
 				Vector2 b = sr[i * 2 + 1];
-				recover_motion += (b - a) * 0.4;
+				recover_motion += (b - a) / cbk.amount;
```
Also, it's a bit outside of this PR's scope, but you could possibly change the same code in `test_body_ray_separation` for consistency reasons (it uses 0.4 too). Would it make sense to you? I'm not sure if it's exactly the same purpose but it seems so.

pouleyKetchoupp commented:
For comparison, here are the results on the same tests for KinematicBody2D on other versions of the code.

**3.2.3 stable**

Body angle 0, Platform size 64:
`Test FAILED: size=64.0, angle=195.0, body angle=0.0`
`Test FAILED: size=64.0, angle=210.0, body angle=0.0`
`Test FAILED: size=64.0, angle=225.0, body angle=0.0`
`Test FAILED: size=64.0, angle=315.0, body angle=0.0`
`Test FAILED: size=64.0, angle=330.0, body angle=0.0`
`Test FAILED: size=64.0, angle=345.0, body angle=0.0`

Body angle 45, Platform size 64:
`Test FAILED: size=64.0, angle=195.0, body angle=45.0`
`Test FAILED: size=64.0, angle=345.0, body angle=45.0`

Body angle 45, Platform size 32:
`Test FAILED: size=32.0, angle=195.0, body angle=45.0`
`Test FAILED: size=32.0, angle=210.0, body angle=45.0`
`Test FAILED: size=32.0, angle=225.0, body angle=45.0`
`Test FAILED: size=32.0, angle=330.0, body angle=45.0`
`Test FAILED: size=32.0, angle=345.0, body angle=45.0`

**This PR without the threshold change on normal checks**

Body angle 0, Platform size 64:
`Test FAILED: size=64.0, angle=180.0, body angle=0.0`

Body angle 45, Platform size 64:
`Test FAILED: size=64.0, angle=180.0, body angle=45.0`
`Test FAILED: size=64.0, angle=195.0, body angle=45.0`
`Test FAILED: size=64.0, angle=345.0, body angle=45.0`

Body angle 45, Platform size 32:
`Test FAILED: size=32.0, angle=15.0, body angle=45.0`
`Test FAILED: size=32.0, angle=30.0, body angle=45.0`
`Test FAILED: size=32.0, angle=150.0, body angle=45.0`
`Test FAILED: size=32.0, angle=165.0, body angle=45.0`
`Test FAILED: size=32.0, angle=180.0, body angle=45.0`
`Test FAILED: size=32.0, angle=195.0, body angle=45.0`
`Test FAILED: size=32.0, angle=210.0, body angle=45.0`
`Test FAILED: size=32.0, angle=225.0, body angle=45.0`
`Test FAILED: size=32.0, angle=330.0, body angle=45.0`
`Test FAILED: size=32.0, angle=345.0, body angle=45.0`

madmiraal commented:
Updated to include:
1. Tolerances for dot product direction calculations. Note: I've used `CMP_EPSILON` to avoid creating another hard-coded constant.
2. Applied the division by `cbk.amount` for recovery motion to `test_body_ray_separation` and to both in Godot 3D too for consistency.

pouleyKetchoupp reviewed (approved):
>     1. Note: I've used `CMP_EPSILON` to avoid creating another hard-coded constant.

That works! It's a lower tolerance value, but the same tests pass so it seems fine.

Given this PR is now fully tested and doesn't produce any obvious regression, I think it's finally ready to go :)

akien-mga commented:
Thanks a ton to everyone involved!

pouleyKetchoupp commented:
@Rhathe Please test your use cases again to see if some of them are still a problem. If it's the case, I would encourage you to open new issues, clearly stating what is your use case and the expected behavior.

https://github.com/godotengine/godot/pull/42574#issuecomment-729754425
https://github.com/godotengine/godot/files/5561017/test_move_one_way_rotated.zip
This case now goes through like in 3.2.3 stable, which is the expected behavior.

https://github.com/godotengine/godot/pull/42574#issuecomment-735468192
https://github.com/godotengine/godot/files/5612930/test_move_one_way_diamond.zip
This case now collides, but it seems like the expected behavior since the body is hitting the top right corner of a one-way collision. This is one of the cases this PR fixes.

chucklepie commented:
Will this fix all jittery issues (fluctuating velocity.y) with using move_and_slide on tilemaps, i.e. with this patch can we safely stop having to put multiple raycasts on our nodes because is_on_floor(), etc fails to return the correct value every other frame?

Thanks in anticipation of good news :)

pouleyKetchoupp commented:
@chucklepie This PR fixes some visible jittering issues, but it's difficult to guarantee it will fix any case of is_on_floor() being inconsistent. Feel free to point out a specific issue or open one with your use case and we'll check if it's already fixed.
